[dogstatsd] readability improvements

remeh opened the pull request:
### What does this PR do?

A few changes are contained in this PR, for review, it's probably better to review the first 5th commit one by one, and to then take a look to the whole PR.
Changes:
  - the metric sample pool and the packet pool have been moved to be global; it's easier to read than having them passed by constructors and object members.
  - instead of passing multiple channels to the batcher, pass the aggregator, simplifying the server and batcher structure + readability of some struct members
  - use the correct type in the UDP listener (instead of using type asserts)

### Motivation

Improve the readability

### Additional Notes

It should have no impact on the Dogstatsd performances.

remeh committed:
dogstatsd: use a global pool for metrics sample and packet buffer instead of a pool declared and shared between all structs.

This has no impact on performances, it is just a matter of claryfing the
code.

Also renamed the PacketBuffer to PacketAssembler in order to avoid confusion
with the PacketsBuffer.

remeh committed:
[aggregator+dogstatsd] Pass the aggregator to the dogstatsd batcher instead of confusing channels.

remeh committed:
[dogstatsd]: properly use a UDPConn in UDP listener.

remeh committed:
[dogstatsd] add a comment.

remeh committed:
Merge branch 'master' into remeh/dogstatsd-readability-improvements

remeh committed:
[aggregator] some fixes due to the merge and small rebase mistakes.

ogaca-dd commented on pkg/dogstatsd/listeners/listeners.go:
```diff
@@ -0,0 +1,7 @@
+
+// GlobalPacketPool is used by the packet assembler to retrieve already allocated
+// buffer in order to avoid allocation.
+var GlobalPacketPool *PacketPool = NewPacketPool(config.Datadog.GetInt("dogstatsd_buffer_size"))
```
Are we confident this code cannot be called before `dogstatsd_buffer_size` is set?

olivielpeau replied:
➕ I don't think this will work as expected, `config.Datadog` is not fully initialized when this is called.

remeh replied:
Good point 🤔 I'll look into that

ogaca-dd commented on pkg/dogstatsd/server_test.go:
```diff
@@ -386,10 +403,13 @@ func TestExtraTags(t *testing.T) {
-	eventOut := make(chan []*metrics.Event)
-	serviceOut := make(chan []*metrics.ServiceCheck)
-	s, err := NewServer(metricOut, eventOut, serviceOut)
+	agg := aggregator.NewBufferedAggregator(
```
What do you think about using `mockAggregator()` instead?
This comment could be applied to the whole file.

ogaca-dd reviewed (commented):
LGTM, just a few nits.

olivielpeau commented on pkg/dogstatsd/server.go:
```diff
@@ -54,15 +55,15 @@ func init() {
+	// listeners are the instanciated socket listener (UDS or UDP or both)
+	listeners []listeners.StatsdListener
+	// aggregator is a pointer to the aggregator that the dogstatsd daemon
+	// will send the the metrics samples, events and service checks to.
```
```suggestion
	// will send the metrics samples, events and service checks to.
```

olivielpeau commented on pkg/dogstatsd/server.go:
```diff
@@ -54,15 +55,15 @@ func init() {
 // Server represent a Dogstatsd server
 type Server struct {
-	listeners                 []listeners.StatsdListener
+	// listeners are the instanciated socket listener (UDS or UDP or both)
```
```suggestion
	// listeners are the instantiated socket listener (UDS or UDP or both)
```

olivielpeau reviewed (commented):
thanks!

Most of these changes are more than welcome! One thing though: I'd avoid global variables whenever possible:
* can we make the `metricPool` an attribute of the aggregator instead?
* can we continue passing the `packetPool` to the structs using it?

I think the lifecycle of an item of the `packetPool` could be made clearer in the code (through comments at least), especially the differences there between udp and uds. Would make the code clearer without having to use global variables.

remeh committed:
Apply suggestions from code review

Co-Authored-By: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

remeh committed:
[dogstatsd] use the mockAggregator in tests where possible.

remeh committed:
Merge branch 'remeh/dogstatsd-readability-improvements' of github.com:DataDog/datadog-agent into remeh/dogstatsd-readability-improvements

remeh committed:
[dogstatsd] do not use global pools.

remeh committed:
[dogstatsd] fix unit tests following changes

remeh commented:
@olivielpeau @ogaca-dd 
I've updated the PR to get rid of the global pools. The metric sample pool is now stored by the aggregator and the packet pool is stored by the listeners. In order to still have a benefit about that from this PR, I've renamed the packet pool to `sharedPacketPool` to be more explicit about its usage.

olivielpeau commented on pkg/dogstatsd/batch.go:
```diff
@@ -18,12 +18,15 @@ type batcher struct {
+	metricSamplePool *metrics.MetricSamplePool
 }
 
 func newBatcher(aggregator *aggregator.BufferedAggregator) *batcher {
```
It'd be nice to avoid shadowing the `aggregator` package with the `aggregator` variable, i.e.:

```suggestion
func newBatcher(agg *aggregator.BufferedAggregator) *batcher {
```

wdyt?

(note: I'm surprised one of our linters doesn't check for this, but this might be common in Go code...)

olivielpeau commented on pkg/dogstatsd/listeners/packet_assembler.go:
```diff
@@ -18,18 +18,22 @@ type packetAssembler struct {
-		flushTimer:    time.NewTicker(flushTimer),
-		closeChannel:  make(chan struct{}),
+		// retrieve an available packet from the packet pool,
+		// the will be pushed by back the server when processed.
```
```suggestion
		// which will be pushed back by the server when processed.
```

olivielpeau commented on pkg/dogstatsd/listeners/packet_assembler.go:
```diff
@@ -69,7 +73,9 @@ func (p *packetAssembler) flush() {
 	p.packetsBuffer.append(p.packet)
-	p.packet = GlobalPacketPool.Get()
+	// retrieve an available packet from the packet pool,
+	// the will be pushed by back the server when processed.
```
```suggestion
	// which will be pushed back by the server when processed.
```

olivielpeau commented on pkg/aggregator/aggregator.go:
```diff
@@ -220,6 +224,8 @@ func NewBufferedAggregator(s serializer.MetricSerializer, hostname, agentName st
 		checkMetricIn:          make(chan senderMetricSample, bufferSize),
 		checkHistogramBucketIn: make(chan senderHistogramBucket, bufferSize),
 
+		MetricSamplePool: metrics.NewMetricSamplePool(32),
```
`32` should be a constant

olivielpeau commented on pkg/dogstatsd/listeners/udp_test.go:
```diff
@@ -19,10 +19,8 @@ import (
-
 func TestNewUDPListener(t *testing.T) {
-	s, err := NewUDPListener(nil, packetPoolUDP)
+	s, err := NewUDPListener(nil, NewPacketPool(32))
```
why `32` instead of `config.Datadog.GetInt("dogstatsd_buffer_size")`? 🤔

olivielpeau commented on pkg/dogstatsd/listeners/packet_assembler_test.go:
```diff
@@ -53,8 +66,9 @@ func TestPacketBufferMergeMaxSize(t *testing.T) {
-	pb, out := buildPacketBuffer(16)
-	message1 := []byte("12345678")
+	pb, out := buildPacketAssembler()
+	// generate a message exactly of the size of the buffer
```
the size of which buffer? 🤔 (the comment could be more precise I guess)

olivielpeau commented on pkg/dogstatsd/listeners/uds_common.go:
```diff
@@ -130,7 +130,9 @@ func (l *UDSListener) Listen() {
 		var err error
-		packet := l.packetPool.Get()
+		// retrieve an available packet from the packet pool,
+		// the will be pushed by back the server when processed.
```
```suggestion
		// which will be pushed back by the server when processed.
```

olivielpeau commented on pkg/dogstatsd/listeners/uds_common_test.go:
```diff
@@ -22,13 +22,11 @@ import (
 	assert.Nil(t, err)
 	defer os.Remove(socketPath)
-	_, err = NewUDSListener(nil, packetPoolUDS)
+	_, err = NewUDSListener(nil, NewPacketPool(512))
```
same question here

olivielpeau commented on pkg/dogstatsd/server.go:
```diff
@@ -250,7 +257,11 @@ func (s *Server) forwarder(fcon net.Conn, packetsChannel chan listeners.Packets)
-	batcher := newBatcher(s.samplePool, s.samplesOut, s.eventsOut, s.servicesCheckOut)
+	// the batcher will be responsible of batching a few samples / events / service
+	// checks and it will automatically forward them to the aggregator, meaning that
+	// the flushing logic to the aggreagtor is actually in the batcher.
```
```suggestion
	// the flushing logic to the aggregator is actually in the batcher.
```

olivielpeau reviewed (commented):
LGTM, a few nits (mostly have questions on the default values used in tests)

remeh committed:
Apply suggestions from code review

Co-Authored-By: Olivier Vielpeau <olivielpeau@users.noreply.github.com>

remeh committed:
[dogstatsd] address feedbacks.

remeh committed:
Merge branch 'remeh/dogstatsd-readability-improvements' of github.com:DataDog/datadog-agent into remeh/dogstatsd-readability-improvements

ogaca-dd reviewed (approved):
LGTM

remeh commented:
@olivielpeau do you want to give it one last look? I've checked the 🔴 CI jobs are unrelated to this

olivielpeau reviewed (approved):
✨
